### DISCUSSION TITLE
Help a beginner

### DISCUSSION THREAD

**Discussion by mugginsjm** [zwave-js/node-red-contrib-zwave-js#153] [Q&A]
I'm sure it is covered here but I can't find it. I have tried many variations of filters but cannot understand how to deal with the "complex" json responses. For example how do I pull out just one piece of data, say temperature or luminance from a multilevel sensor so that I can use this value to make a logic decision. I have experimented with Json manipulation in a function but I'm sure I'm missing something really obvious. 
Some help really appreciated.

**Comment by marcus-j-davies** [comment]
Hi @mugginsjm,

And welcome to the project.

So..

On the basis you want to respond to temp changes, luminance changes etc. etc.

1 -  bring in a filter node (attach it to the right hand side of the controller (or a device node)
![image](https://user-images.githubusercontent.com/55892693/142385135-f1e536d0-a40e-4f7e-a240-1609ac64234c.png)

Double clicking this node, you can add 'rules' to it.
A rule states what value updates this rule will output for, so if you have 2 rules, it will create 2 outputs.
![image](https://user-images.githubusercontent.com/55892693/142386483-ec570014-80c7-4b33-aa9f-a0458223f747.png)

So how do you tell each rule to only listen for specific value updates?
Activate a filter, by clicking it (it will expand).

2 -  On the Zwave UI Tab, double click a values title, and you will get the option to add it to  the filter you have activated.
![image](https://user-images.githubusercontent.com/55892693/142387394-740f70f9-f0f5-4d3a-83ae-f534fbddf4b0.png)

The end Result
![image](https://user-images.githubusercontent.com/55892693/142388220-37e082dd-aaad-4874-85ed-116ccf48ac1f.png)

The value that you want to process in you function will normally be in `msg.payload.object`

**Reply by marcus-j-davies** [reply]
We do have a getting started flow - which covers quite a lot.
the Wiki is also very detailed.

https://github.com/zwave-js/node-red-contrib-zwave-js/wiki/First-Z-Wave-Flow-Setup

**Reply by mugginsjm** [reply]
Marcus thank you for the response. I have already (I think) got to grips with the filter. My problem is extracting a specific item eg temperature fron the payload.object.
John

**Reply by marcus-j-davies** [reply]
No worries,

Every Command Class Type will deliver their events via an `object`
this `object` will detail the specific Command Class it refers to and the endpoint of the source node.

Here is the event for a Mulitlevel Sensor change

```javascript
{
  "payload": {
    "node": 10,
    "event": "VALUE_UPDATED",
    "object": {
      "commandClassName": "Multilevel Sensor",
      "commandClass": 49,
      "endpoint": 0,
      "property": "Air temperature",
      "propertyName": "Air temperature",
      "newValue": 24
    },
    "timestamp": 1637014660258,
  }
}
```

We can see here, there is a very interesting property called `newValue`. (it can differ across CC)
this will be the value we care about.

based on the fact this passes through a filter node, you only need to worry about `newValue `- without having to worry, if this is a temp or luminance value change - the filter will direct them out the correct output.

Knowing this:, you could do this in a function node
```javascript
if (msg.payload.object.newValue < 18) {
    // turn on heating on
} else {
    // turn heating off
}
```

The overall events will be:
```javascript
{
  "payload": {
    "node": ##,
    "event": "SOME_EVENT",
    "object": {},
    "timestamp": 1637014660258,
  }
}
```

anything passing through a filter will also add `msg.filter`

**Reply by mugginsjm** [reply]
Marcus you're a superstar. The "msg.payload.object.newValue" is what I needed.
I had read all the wiki and associated notes and it is very compehensive. It seems that most contributors have no problem with all the vagaries of payload objects etc. Up until now I have only been dealing with simple payloads.
It might be useful to others like me to add a tiny section to the wiki, showing maybe just this line of code.
thanks again.

**Reply by marcus-j-davies** [reply]
@crxporter,

When your ready 😉 

Maybe put something in the wiki, on how to actually get to the value that got updated.
I think maybe a couple  of examples, like the below?
```javascript
if (msg.payload.object.newValue < 18) {
    // turn heating on
} else {
    // turn heating off
}
```

Multilevel Sensor, Binary Sensor, use `newValue`, there could be others that use different names for the new value, Like Binary Switch, or Basic Switch

I'll leave that in your capable hands

**Reply by crxporter** [reply]
> Maybe put something in the wiki, on how to actually get to the value that got updated.
I think maybe a couple of examples, like the below?

Seems like an excellent idea. I got so caught up in teaching the Z-wave parts that I forgot some people may not be so familiar with finding the important data in a JSON !!

Probably should show it like a switch node as well. Temp < 18 goes one way, greater another way. Also works nice with true/false on a switch... and so on.
